docs/src/user_guide/setup/highlighting.md

```diff
@@ -0,0 +1,133 @@
+
+Now we have to update the card templates and styling.
+
+## Changing Your Card Templates
+
+Here is a simplified template of the card shown above without any highlighting:
+
+![front-template-basic.png](../../img/front-template-basic.png)
+
+![back-template-basic.png](../../img/back-template-basic.png)
+
+
+To add **dynamic** highlighting to the back of the card we prepend `am-highlight:` to the `Japanese` field:
+![back-template-dynamic-highlighting.png](../../img/back-template-dynamic-highlighting.png)
+
+To add **static** highlighting to the back of the card we replace the `Japanese` field with `am-highlighted`
+![back-template-static-highlighting.png](../../img/back-template-static-highlighting.png)
+
+
+We also need to update the `styling` section to specify which colors we want the morphs to have.
+
+## Changing Your Card Styling
+
+![styling.png](../../../img/styling.png)
+
+You can pick and choose among these; if you only want unknown morphs to be highlighted, and you don't care about
+dark-mode, then only adding the first line would be enough. You can also change the colors to anything you want.
+
+
+``` css
+[morph-status=unknown] { color: blue; }
+[morph-status=learning] { color: #8bb33d; } /* light-green */
+[morph-status=known] { color: green; }
+
+.nightMode [morph-status=unknown] { color: red; } 
+.nightMode [morph-status=learning] { color: #ffff99; } /* yellow */
+.nightMode [morph-status=known] { color: #8bb33d; } /* light-green */
+```
+
+
+
+It’s also possible to use `background-color`:
+
+``` css
+[morph-status=unknown] { background-color: #f7867e; } /* red */
+[morph-status=learning] { background-color: #ffff99; } /* yellow */
+[morph-status=known] { background-color: #49f53e; } /* green */
+
+.nightMode [morph-status=unknown] { background-color: #b74d4d; } /* red */
+.nightMode [morph-status=learning] { background-color: #ccad50; } /* yellow */
+.nightMode [morph-status=known] { background-color: #27961f; } /* green */
+```
+
+
+### Furigana and other ruby characters
+
+![ruby_characters_preserved.png](../../../img/ruby_characters_preserved.png)
+
+> **Note**: The `Ignore content in square brackets` [preprocess setting](../setup/settings/preprocess.md) option needs to
+> be activated for ruby character highlighting to function properly.
+
+To have [ruby characters](https://docs.ankiweb.net/templates/fields.html#ruby-characters) such as furigana displayed,
+you have to prepend the respective [character filter](https://docs.ankiweb.net/templates/fields.html#additional-ruby-character-filters)
+to the field on the card
+template, e.g:
+
+``` text
+{{furigana:am-highlighted}}
+```
+or
+
+``` text
+{{am-highlight:furigana:Japanese}}
+```
+
+
+> **Note**: Because [Anki cannot process custom filters before built in filters](https://github.com/ankicommunity/anki-desktop/blob/main/rslib/src/template_filters.rs#L22-L24),
+> it is required to apply the character filter to the right of `am-highlight:`
```

Comment:
I think we could add some clarity here about when you need/ want to use furigana.

e.g. if you will always do your reviews on a machine that has ankimorphs, then you don't need furigana in the second example. 

the problem  is that it's hard to explain to the average user all of these possible conditions, and therefore may bloat the docs....

I don't have a good answer, just bringing it up.

Reply:
> e.g. if you will always do your reviews on a machine that has ankimorphs, then you don't need furigana in the second example.

oh wait, it's only necessary for mobile?

Reply:
correct. in the new impl. it does the rubification that furigana does.

the only time you need {{am-highlight:furigana:Japanese}} is as a fallback where am-highligted is not available BUT you still want your furigana. in that case the `furigana:` filter runs, then the `am-highlight:` is a no-op, because the add-on is not available (e.g. on my iOS version I need cards to have that in the template, idk about AnkiDroid, does it have the ability to run plugins?)

in the case where you do have both and `am-highlight` IS available, the user should `am-highlight:Japanese`, as it's EVER so slightly faster JIT.

But if they don't do that, and have them stacked together as `am-highlight:furigana:Japanese` we actually de-construct the rubies (see code below). This way it works on raw field data OR pre-processed `furigana:` data.

```
    ruby_longhand = r"(?:<ruby[^<]*>)(?:<rb[^>]*>|.{0})(?P<base>.*?)(?:</rb>|.{0})<rt[^>]*>(?P<ruby>.+?)</rt>(?P<after>.*?)(?:</ruby>)"

    # Emit the captured ruby into square brackets, thus reconstructing the ruby shorthand "X[yz]".
    # Pad with a leading space so that we can retain the base/ruby relationship
    ruby_shorthand = r" \g<base>[\g<ruby>]\g<after>"

    text = re.sub(ruby_longhand, ruby_shorthand, text, flags=re.IGNORECASE).strip()
```

Reply:
Yeah, that is useful info.

I was going to propose this:

---
> **Note**: The `Ignore content in square brackets` [preprocess setting](../setup/settings/preprocess.md) option needs to  
> be activated for ruby character highlighting to function properly.  

The highlighted text generated by AnkiMorphs will automatically display [ruby characters](https://docs.ankiweb.net/templates/fields.html#ruby-characters) such as furigana.

If you are using dynamic highlighting, you may want to combine it with the `furigana:`  [character filter](https://docs.ankiweb.net/templates/fields.html#additional-ruby-character-filters) as a fallback mechanism. This makes sure that furigana will visible on mobile devices or computers without the add-on. Here's how you would do that:

``` text
{{am-highlight:furigana:Japanese}}  
```

> **Note**: Because [Anki cannot process custom filters before built in filters](https://github.com/ankicommunity/anki-desktop/blob/main/rslib/src/template_filters.rs#L22-L24),  
> it is required to apply the character filter to the right of `am-highlight:`

You can hide furigana by applying the `kanji:` character filter:

  
``` text
{{kanji:am-highlighted}}  
```  
or  
  
``` text
{{am-highlight:kanji:Japanese}}  
```  
---

But I just tried to use `{{kanji:am-highlighted}}` and the furigana are still shown... It works as expected with the `am-highlight` filter though.

Reply:
@RobHelgeson do you think there is an easy workaround, or should we leave it as a known bug?

Reply:
Let me take a look; I THINK it'll not be possible because we're already at the `<ruby>` stage, so the kanji template filter can't find anything that looks like `abc[123]`. 

bit of a crappy workaround but we could add some notes about css workarounds:

```
# this equals the `kanji` filter.
ruby rt {
	visibility: hidden;
}
```

still thinking about a good css fix for the `kana` template filter, probably something like (this does not cover all cases yet)
```
# this equals the `kana` filter.
ruby {
	display: inline-block;

  & span {
    visibility: hidden;
	}

  & rt {
    font-size: 100%;
  }
}
```

Reply:
Here is the kana filter emulated.

```
# this equals the `kana` filter.
ruby {
	display: inline-block;
	visibility: hidden;

  & span {
    visibility: hidden;
	}

  & rt {
    font-size: 100%;
	  visibility: visible;
  }
}
```

Reply:
> it'll not be possible because we're already at the `<ruby>` stage

Aha, so `{{kanji:am-highlighted}}` is equivalent to `{{kanji:am-highlight:Japanese}}`, which won't work.

I tired your css on a text with multiple lines, and the line-height got a little wonky:

![Screenshot from 2024-12-20 14-52-28](https://github.com/user-attachments/assets/77d8829b-8864-423b-866f-a7627a5993ae)

![Screenshot from 2024-12-20 14-12-31](https://github.com/user-attachments/assets/05d9eeda-340d-4c8a-b2f1-a6523a0edacb)


So instead of just hiding the ruby, I changed it to not display, which seems to work better:

![Screenshot from 2024-12-20 14-40-49](https://github.com/user-attachments/assets/20323716-f009-4a11-a3d1-30ca90f6410e)

```css
/* kana: */
ruby * {
  display: none;
}

ruby rt {
  display: inline;
  font-size: inherit;
}

```
```css
/* kanji: */
ruby rt {
    display: none;
}
```

Do you foresee any problems with that that approach?

Reply:
I think the Kana one will not work with `display: none;`, because it will need its children to override `display: none;` and I don't think this can't be done, `display: none;` takes it out of the Dom (essentially), that's why I tried with visibility. 

The Kanji one should be fine.

We might need yours for kanji and mine for kana. I just realized that the kana selector should be wrapped in a class. If it's present but the user does not always want to use it, it will break the rubies in the way you show, because it changes ALL rubies to `display: inline-block;` which is 99% not what is wanted.

I'll play around with this a little right now.

Reply:
> I think the Kana one will not work with display: none;, because it will need its children to override display: none; and I don't think this can't be done, display: none; takes it out of the Dom (essentially), that's why I tried with visibility.

It uses `ruby * {}` not `ruby {}`, It produced the last picture that only has kana.

Reply:
How about this? I think it covers all the cases.

`<div>{{am-highlight:furigana:Japanese}}</div>` or `{{am-highlight:furigana:Japanese}}`

`<div class="kana">{{am-highlight:kana:Japanese}}</div>`

`<div class="kanji">{{am-highlight: kanji:Japanese}}</div>`

Accompanying css for the last 2.

``` css
# Within the kana class tagged div,
# rubies
# set the visibility of the rubies to invisible so they are not shown
# set the display to inline-block to collapse the rt's inline with the rubies.
# rt's
# make the rt's have the same font size as the rubies do (by browser default they are ~65%)
# move them up so that they're not causing the whole ruby to be 2x height (a little hacky)
# set them to visible
.kana {
  & ruby {
    display: inline-block;
    visibility: hidden;
  
    & rt {
      font-size: inherit;
      margin-top: -1em;
      visibility: visible;
    }
  }
}
```

``` css
# Completely remove rt's from the DOM.
.kanji {
  & ruby rt {
      display: none;
  }
}
```

I will say, having the users do this seems a bit..... meh.... I suppose we could have 3 different filters, that would solve this quite easily.

Reply:
> > I think the Kana one will not work with display: none;, because it will need its children to override display: none; and I don't think this can't be done, display: none; takes it out of the Dom (essentially), that's why I tried with visibility.
> 
> It uses `ruby * {}` not `ruby {}`, so the element itself is still there. It produced the last picture that only has kana, so the overriding seems to work.

oh, this inlines the kanji and the kana, my version supresses the kanji and allows only kana.

I need to look and see what the Anki built in pipe does!

Reply:
Ah, nvm, my approach only worked on that one example:
```css
<span style="white-space: pre-wrap;"><ruby><span morph-status="unknown">錬金術</span><span morph-status="unknown">師</span><rt>れんきんじゅつし</rt></ruby></span>
```

on this it completely fails:
```css
<span morph-status="known">お</span><span morph-status="unknown"><ruby>疲れ様<rt>つかれさま</rt></ruby>でした</span>
```

Reply:
do you know why the first example has the ruby wrapped around the span, but the second example doesn't?

Reply:
yes, if the morph === the ruby (or completely contains the ruby) then span on the outside, but if there are more than one morph inside the ruby each COULD have a different learning step and we must color them independently so the ruby has to go on the outside.. 

Here are all the different combos.
```
 |-mmm---| |--mmm--| |---mmm-| |-mmm---| |-mmmmm-| |-mmmmm-| |--mmm--| |--m-m--|
 |----rrr| |--rrr--| |--rrr--| |--rrr--| |--rrr--| |--r-r--| |-rrrrr-| |-rrrrr-|
```

These represent all of the different cases in the main process method.

Reply:
1. No overlap
2. same overlap (span goes on the outside)
3. Ruby starts, then morph starts, ruby ends, morph ends
4. same but in reverse
5. morph completely wraps the ruby ---- YOUR EXAMPLE #2
6. (specialized case) morph completely wraps more than one ruby
7. same as 5 in reverse
8. same as 6 in reverse ---- YOUR EXAMPLE #1